cmd/cstor-pool-mgmt/controller/replica-controller/handler.go

```diff
@@ -0,0 +1,208 @@
+/*
+Copyright 2018 The OpenEBS Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package replicacontroller
+
+import (
+	"fmt"
+	"os"
+	"reflect"
+
+	"github.com/golang/glog"
+	"github.com/openebs/maya/cmd/cstor-pool-mgmt/controller/common"
+	"github.com/openebs/maya/cmd/cstor-pool-mgmt/volumereplica"
+	apis "github.com/openebs/maya/pkg/apis/openebs.io/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/tools/cache"
+)
+
+// syncHandler compares the actual state with the desired, and attempts to
+// converge the two. It then updates the Status block of the CStorReplicaUpdated resource
+// with the current status of the resource.
+func (c *CStorVolumeReplicaController) syncHandler(key, operation string) error {
+	cVRGot, err := c.getVolumeReplicaResource(key, operation)
+	if err != nil {
+		return err
+	}
+	status, err := c.cVREventHandler(operation, cVRGot)
+	if status == "" {
+		return nil
+	}
+	cVRGot.Status.Phase = apis.CStorVolumeReplicaPhase(status)
+	if err != nil {
+		glog.Errorf(err.Error())
+		glog.Infof("cVR:%v, %v; Status: %v", cVRGot.Name,
+			string(cVRGot.GetUID()), cVRGot.Status.Phase)
+		_, err := c.clientset.OpenebsV1alpha1().CStorVolumeReplicas().Update(cVRGot)
+		if err != nil {
+			return err
+		}
+		return err
+	}
+	_, err = c.clientset.OpenebsV1alpha1().CStorVolumeReplicas().Update(cVRGot)
+	if err != nil {
+		return err
+	}
+	glog.Infof("cVR:%v, %v; Status: %v", cVRGot.Name,
+		string(cVRGot.GetUID()), cVRGot.Status.Phase)
+	return nil
+
+}
+
+func (c *CStorVolumeReplicaController) cVREventHandler(operation string, cVR *apis.CStorVolumeReplica) (string, error) {
+	err := volumereplica.CheckValidVolumeReplica(cVR)
+	if err != nil {
+		c.recorder.Event(cVR, corev1.EventTypeWarning, common.FailureValidate, common.MessageResourceFailValidate)
+		return string(apis.CVRStatusOffline), err
+	}
+
+	// PoolNameHandler tries to get pool name and blocks for
+	// particular number of attempts.
+	var noOfAttempts = 5
+	if !common.PoolNameHandler(cVR, noOfAttempts) {
```

Comment:
Since we should not be starting the volumereplica controller unless the cstor pool is created/imported, we need not check multiple times for the pool name. TODO: Add more checks for the status of the pool object, like the volume request should not be processed if the pool is in a non-serviceable state (offline/faulted/unavailable).

Reply:
Changed noOfAttempts to 2. As discussed, we can consider and handle additional status in next design phase/PR.